src/adc.rs

```diff
@@ -0,0 +1,562 @@
+//! # API for the Analog to Digital converter
+
+#![allow(dead_code)]
+
+use crate::rcc::{Clocks, APB2};
+
+use crate::gpio::Analog;
+use crate::gpio::{gpioa, gpiob, gpioc, gpiof};
+
+use crate::pac::{ADC1, ADC2, ADC3, ADC_COMMON};
+
+use cortex_m::asm::delay;
+
+use embedded_hal::adc::{Channel, OneShot};
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(non_camel_case_types)]
+/// ADC sampling time
+///
+/// Options for the sampling time, each is T ADC clock cycles.
+// 15.13.4 >> ADC sample time register
+pub enum SampleTime {
+    /// 3 cycles sampling time
+    T_3,
+    /// 15 cycles sampling time
+    T_15,
+    /// 28 cycles sampling time
+    T_28,
+    /// 56 cycles sampling time
+    T_56,
+    /// 84 cycles sampling time
+    T_84,
+    /// 112 cycles sampling time
+    T_112,
+    /// 144 cycles sampling time
+    T_144,
+    /// 480 cycles sampling time
+    T_480,
+}
+
+impl Default for SampleTime {
+    /// Get the default sample time (currently 56 cycles)
+    fn default() -> Self {
+        SampleTime::T_56
+    }
+}
+
+impl From<SampleTime> for u8 {
+    fn from(val: SampleTime) -> Self {
+        use SampleTime::*;
+        match val {
+            T_3 => 0,
+            T_15 => 1,
+            T_28 => 2,
+            T_56 => 3,
+            T_84 => 4,
+            T_112 => 5,
+            T_144 => 6,
+            T_480 => 7,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+/// ADC data register alignment
+pub enum Align {
+    /// Right alignment of output data
+    Right,
+    /// Left alignment of output data
+    Left,
+}
+
+impl Default for Align {
+    /// Default: right alignment
+    fn default() -> Self {
+        Align::Right
+    }
+}
+
+// 15.13.3 ADC control register >> Bit 11 ALIGN: Data alignment
+impl From<Align> for bool {
+    fn from(val: Align) -> Self {
+        match val {
+            Align::Right => false,
+            Align::Left => true,
+        }
+    }
+}
+
+/////////////////////////////////
+
+macro_rules! adc_pins {
+    ($ADC:ident, $($pin:ty => $chan:expr),+ $(,)*) => {
+        $(
+            impl Channel<$ADC> for $pin {
+                type ID = u8;
+
+                fn channel() -> u8 { $chan }
+            }
+        )+
+    };
+}
+
+// See "Datasheet - production data"
+// Pinouts and pin description (page 66..)
+adc_pins!(ADC1,
+    gpioa::PA0<Analog> => 0_u8,
```

Comment:
Since the type is known within the macro-generated code, it's probably unnecessary to have the `_u8` on all these numbers. Doesn't really matter though.